quantum/rgb_matrix.c

```diff
@@ -539,356 +423,328 @@ void rgb_matrix_digital_rain( const bool initialize ) {
         toggle_enable_last = rgb_matrix_config.enable;
-    #endif
-
-    // this gets ticked at 20 Hz.
-    // each effect can opt to do calculations
-    // and/or request PWM buffer updates.
-    switch ( effect ) {
-        case RGB_MATRIX_SOLID_COLOR:
-            rgb_matrix_solid_color();
-            break;
-        #ifndef DISABLE_RGB_MATRIX_ALPHAS_MODS
-            case RGB_MATRIX_ALPHAS_MODS:
-                rgb_matrix_alphas_mods();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_DUAL_BEACON
-            case RGB_MATRIX_DUAL_BEACON:
-                rgb_matrix_dual_beacon();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_GRADIENT_UP_DOWN
-            case RGB_MATRIX_GRADIENT_UP_DOWN:
-                rgb_matrix_gradient_up_down();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_RAINDROPS
-            case RGB_MATRIX_RAINDROPS:
-                rgb_matrix_raindrops( initialize );
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_CYCLE_ALL
-            case RGB_MATRIX_CYCLE_ALL:
-                rgb_matrix_cycle_all();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_CYCLE_LEFT_RIGHT
-            case RGB_MATRIX_CYCLE_LEFT_RIGHT:
-                rgb_matrix_cycle_left_right();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_CYCLE_UP_DOWN
-            case RGB_MATRIX_CYCLE_UP_DOWN:
-                rgb_matrix_cycle_up_down();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_RAINBOW_BEACON
-            case RGB_MATRIX_RAINBOW_BEACON:
-                rgb_matrix_rainbow_beacon();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_RAINBOW_PINWHEELS
-            case RGB_MATRIX_RAINBOW_PINWHEELS:
-                rgb_matrix_rainbow_pinwheels();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_RAINBOW_MOVING_CHEVRON
-            case RGB_MATRIX_RAINBOW_MOVING_CHEVRON:
-                rgb_matrix_rainbow_moving_chevron();
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_JELLYBEAN_RAINDROPS
-            case RGB_MATRIX_JELLYBEAN_RAINDROPS:
-                rgb_matrix_jellybean_raindrops( initialize );
-                break;
-        #endif
-        #ifndef DISABLE_RGB_MATRIX_DIGITAL_RAIN
-            case RGB_MATRIX_DIGITAL_RAIN:
-                rgb_matrix_digital_rain( initialize );
+
+        if (initialize) {
+            // wipe all the LEDs, since some effects never touch underglow leds,
+            // resulting in a static-underglow when there shouldn't be any
+            for (int i = 0; i < DRIVER_LED_TOTAL; i++)
+                rgb_matrix_all_off(i);
+        }
+    }
+
+    // TODO?: tweak effects and timings to works with nonstandard #leds per run
+    #define LED_PER_RUN DRIVER_LED_TOTAL
+
+    for (int i = 0; i < LED_PER_RUN; i++) {
```

Comment:
I've got some serious concerns with the changes to rgb_matrix.c. You took the loop out of each animation function and placed it here instead. This has a very nasty performance side effect of putting all the math that was removed from loops in the animation functions right back into the loop. Even worse is that now you are looping over a giant branch statement, and calling functions pushing and popping data off the stack. This really needs to be undone.

Reply:
A fair point! But this apparent trade in performance offers up quite a few solid benefits:
- It allows easy composition of multiple effects on a per-led level. In the future, it might even be possible to extend the QMK configurator to allow for some sort of easy per-key lighting customization!
- The initial reason for pulling the for-loop out was because the initial Massdrop code batched LED updates, instead of how RGB Matrix did it, by having them all fire off at once. That's actually why there is that seemingly useless #define there, as it should be possible to set it to a smaller number, and enable batched LED updates. In the end, the other improvements to the system resulted in not having to enable this functionality on the CTRL/ALT, but hey, it's nice to have, in case an under-powered board wants to leave more time for other tasks, but still have lighting effects.
- Having a single loop is cleaner from a software-engineering point of view (IMHO)

As for the performance considerations, I have 2 responses
- I've wrapped particularly expensive operations (i.e: the ones that used to be outside of the for loop in each effect) in a `if (led_i == 0)` block, and marked the variables as static, so they are only computed once-per-frame.
- If you're worried about looping through a switch, it's pretty easy to just set a function pointer to the desired effect per-frame outside the loop. That should alleviate any serious performance penalty from having to traverse the switch statement (though if the switch is optimized into a jump table, it shouldn't be too bad to begin with (idk though, haven't checked myself))

Reply:
Those benefits sound great and all, but they are still achievable without hoisting the loop, moving variables into global scope, introducing more branching logic, and so on. This approach has very real performance impacts on AVR (something I've been working on getting working) and impact on ARM to a lesser extent. So lets work together and break up this pr into more manageable chunks.